packages/app-builder-lib/src/node-module-collector/index.ts

```diff
@@ -3,21 +3,47 @@ import { PnpmNodeModulesCollector } from "./pnpmNodeModulesCollector"
 import { YarnNodeModulesCollector } from "./yarnNodeModulesCollector"
 import { detect, PM, getPackageManagerVersion } from "./packageManager"
 import { NodeModuleInfo } from "./types"
+import { readFile } from "fs-extra"
+import * as path from "path"
+import { exists } from "builder-util"
 
 async function getCollectorByPackageManager(rootDir: string) {
   const manager: PM = await detect({ cwd: rootDir })
+  const npmrc = await parseNpmrc(rootDir)
   switch (manager) {
-    case "npm":
-      return new NpmNodeModulesCollector(rootDir)
     case "pnpm":
+      if (npmrc?.["node-linker"] === "hoisted") {
+        return new NpmNodeModulesCollector(rootDir)
+      }
       return new PnpmNodeModulesCollector(rootDir)
+    case "npm":
+      return new NpmNodeModulesCollector(rootDir)
     case "yarn":
       return new YarnNodeModulesCollector(rootDir)
     default:
       return new NpmNodeModulesCollector(rootDir)
   }
 }
 
+async function parseNpmrc(rootDir: string) {
```

Comment:
We should preferably use `pnpm config get node-linker`, as it works even in the case of two `package.json` files or within a workspace.

![image](https://github.com/user-attachments/assets/33ccd829-e4c2-4023-b3a6-0059160a0844)

Reply:
Refers to https://github.com/electron-userland/electron-builder/pull/8886

Reply:
That's fantastic. Definitely will use that.

Reply:
Since we need to check `shamefully-hoist=true` `public-hoist-pattern=*` and `node-linker=hoisted`, `pnpm config list` with a `reduce` to `Record<string, string>` should do the trick.

Not sure if we can easily check variations of `public-hoist-pattern` though, I have no clue how to parse it atm
https://pnpm.io/npmrc#public-hoist-pattern

Reply:
Under these two configurations(`shamefully-hoist=true public-hoist-pattern=*`), we can still use pnpm list instead of npm list.